dags/synapse-dataset-to-croissant.py

```diff
@@ -165,17 +199,559 @@ def set_up_logging() -> Logger:
+    return distribution_files
+
+
+def construct_record_set_section_for_files(files_attached_to_dataset: List[File]) -> Dict[str, str]:
+    """
+    Construct the record set section for the files attached to the dataset. This is used
+    to extract the keys from the annotations of the files and create a Field for each
+    key.
+
+    Arguments:
+        files_attached_to_dataset: The list of files attached to the dataset.
+
+    Returns:
+        The record set section for the files attached to the dataset.
+    """
+    unique_annotation_keys = set()
+
+    for file in files_attached_to_dataset:
+        file: File = file
+        file_annotations = file.annotations.keys()
+        for key in file_annotations:
+            unique_annotation_keys.add(key)
+
+    metadata_fields = []
+    # Sort the set of unique_annotation_keys ignoring case, but keep the case in the result
+    unique_annotation_keys = sorted(unique_annotation_keys, key=str.casefold)
+
+    for key in unique_annotation_keys:
+        metadata_fields.append(
+            {
+                "@type": "Field",
+                "@id": f"metadata/{key}",
+                "name": key,
+                "description": "",
+                "dataType": "sc:Text",
+                "source": {
+                    "fileObject": {"@id": "metadata"},
+                    "extract": {"column": key}
+                }
+            }
+        )
+
+    return {
+        "@type": "RecordSet",
+        "@id": "default",
+        "name": "default",
+        "description": "Metadata for the dataset",
+        "field": metadata_fields
+    }
+
+
+def extract_s3_objects_to_delete(bucket_objects: List[str], dataset_collections_to_consider_for_deletion: List[str], combined_dataset_collection_and_datasets: List[Dict[str, str]]) -> List[str]:
+    """
+    Extract the S3 objects to delete from the S3 bucket. This is done by retrieving
+    all of the objects in the S3 bucket and checking if the object matches one of the
+    datasets that will be kept after the full DAG as been run. Every object in the S3
+    bucket that matches the pattern `_syn####.###_` and
+    `_datasetCollection_syn####_` will be considered for deletion. The DAG param
+    `dataset_collections_for_cleanup` is used to filter the objects that will be
+    considered for deletion as the second filter.
+
+    The last filter enforces that each object in the S3 bucket must match a dataset
+    id, dataset version, dataset collection, and dataset name in the
+    `combined_dataset_collection_and_datasets` list. If the object does not match
+    any of the objects in the list, then it will be marked for deletion.
+
+    Arguments:
+        bucket_objects: The list of objects in the S3 bucket.
+        dataset_collections_to_consider_for_deletion: The dataset collections to
+            consider for deletion. This is passed in from the
+            `dataset_collections_for_cleanup` DAG parameter.
+    """
+    objects_to_delete = []
```

Comment:
can we also add more comments to this function?

Reply:
Re-read over the docstring of this function and let me know what your thoughts are. The code is relatively straight forward

Reply:
Thanks for updating the docstring. I understand the most parts, but I am curious why it is important to enforce the last filter? why each object in the S3 bucket must match a dataset id, dataset version, dataset collection, and dataset name? is it for ensuring that we delete S3 objects that were no longer present in the dataset collections? I know that you have mentioned that in the PR description, but could you add a comment in the code or clarify in the docstring? I am thinking referencing the PR description might not be as convenient as looking at the code comment.

Reply:
I think I understand now after reading more about the doc string in delete_non_current_croissant_file_in_s3 function. We could keep the code there as it is. Thank you

Reply:
@linglp thanks for the thoughtful review. My 2 cents is that module docstring covers this information on line 12 which is the requirement to remove the json from the S3 bucket for two scenarios:

1. Datasets removed from the dataset collection
2. Datasets made private from public (since these croissant files are publicly accessible)

I'll defer to Bryan if he would like to add more information but imo, we also don't need to over document.